apis/apiextensions/v1/composition_types.go

```diff
@@ -843,3 +830,20 @@ type CompositionList struct {
 	metav1.ListMeta `json:"metadata,omitempty"`
 	Items           []Composition `json:"items"`
 }
+
+// PatchFieldValueToObject applies the value to the "to" object at the given
+// path with the given merge options, returning any errors as they occur.
+// If no merge options is supplied, then destination field is replaced
+// with the given value.
+func PatchFieldValueToObject(fieldPath string, value interface{}, to runtime.Object, mo *xpv1.MergeOptions) error {
+	paved, err := fieldpath.PaveObject(to)
+	if err != nil {
+		return err
+	}
+
+	if err := paved.MergeValue(fieldPath, value, mo); err != nil {
+		return err
+	}
+
+	return runtime.DefaultUnstructuredConverter.FromUnstructured(paved.UnstructuredContent(), to)
+}
```

Comment:
Can this function live outside our API type definition files? i.e. Somewhere under `internal/`? As currently implemented it doesn't appear to actually have any dependency on the API types defined in `package v1`. I'd actually personally prefer for our API type definitions to be as close to plain structs without any logic as possible, but I haven't yet found a good way to do that.

Another option here would be to just keep this function private and duplicate it in `merge.go`.

Reply:
I have duplicated it although, in my opinion, preventing a duplicate is a stronger incentive. Nowhere, except probably the `apiextensions/composite` package (another place this is used) looked suitable for me but I think it would not be suitable to import something from `apiextensions/composite` in `apiextensions/v1`. So I preferred duplication.

Reply:
> I have duplicated it although, in my opinion, preventing a duplicate is a stronger incentive.

Interestingly this is one of many areas where Go takes a bit of an atypical approach - in fact "a little copying is better than a little dependency" is one of the "Go proverbs" per https://go-proverbs.github.io/ and https://www.youtube.com/watch?v=PAAkCSZUG1c&t=568s.